lib/transformer.js

```diff
@@ -0,0 +1,228 @@
+"use strict";
+
+const path = require("path");
+
+const Promise = require("bluebird");
+const fs = require("fs-extra");
+const webidl = require("webidl2");
+
+const Interface = require("./constructs/interface");
+const Dictionary = require("./constructs/dictionary");
+
+class Transformer {
+  constructor() {
+    this.options = { implSuffix: "" };
+    this.sources = [];
+    this._interfaces = null;
+    this._dictionaries = null;
+    this._customTypes = null;
+  }
+
+  addSource(idl, impl) {
+    if (typeof idl !== "string") {
+      throw new TypeError("idl path has to be a string");
+    }
+    if (typeof impl !== "string") {
+      throw new TypeError("impl path has to be a string");
+    }
+    this.sources.push({ idlPath: idl, impl });
+    return this;
+  }
+
+  *_collectSources() {
+    const stats = yield Promise.all(this.sources.map((src) => Promise.fromNode((next) => fs.stat(src.idlPath, next))));
```

Comment:
Or I guess maybe bluebird's promisifyAll is the way to go since you're not using the standard fs module